flink-runtime/src/test/java/org/apache/flink/runtime/source/coordinator/SourceCoordinatorAlignmentTest.java

```diff
@@ -131,6 +136,73 @@ void testWatermarkAlignmentWithTwoGroups() throws Exception {
         }
     }
 
+    /**
+     * When JobManager failover and auto recover job, SourceCoordinator will reset twice: 1. Create
+     * JobMaster --> Create Scheduler --> Create DefaultExecutionGraph --> Init
+     * SourceCoordinator(but will not start it) 2. JobMaster call
+     * restoreLatestCheckpointedStateInternal, which will call {@link
+     * org.apache.flink.runtime.operators.coordination.RecreateOnResetOperatorCoordinator#resetToCheckpoint(long,byte[])}
+     * and reset SourceCoordinator. Because the first SourceCoordinator is not be started, so the
+     * period task can't be stopped.
+     */
+    @Test
+    void testSourceCoordinatorReset() throws Exception {
+        long maxDrift = 1000L;
+        WatermarkAlignmentParams params =
+                new WatermarkAlignmentParams(maxDrift, "group1", maxDrift);
+
+        final Source<Integer, MockSourceSplit, Set<MockSourceSplit>> mockSource =
+                createMockSource();
+
+        // First to init a SourceCoordinator to simulate JobMaster init SourceCoordinator
+        AtomicInteger counter1 = new AtomicInteger(0);
+        sourceCoordinator =
+                new SourceCoordinator<MockSourceSplit, Set<MockSourceSplit>>(
+                        OPERATOR_NAME,
+                        mockSource,
+                        getNewSourceCoordinatorContext(),
+                        new CoordinatorStoreImpl(),
+                        params,
+                        null) {
+                    @Override
+                    void announceCombinedWatermark() {
+                        counter1.incrementAndGet();
+                    }
+                };
+
+        // Second we call SourceCoordinator::close and re-init SourceCoordinator to simulate
+        // RecreateOnResetOperatorCoordinator::resetToCheckpoint
+        sourceCoordinator.close();
+        AtomicInteger counter2 = new AtomicInteger(0);
+        sourceCoordinator =
+                new SourceCoordinator<MockSourceSplit, Set<MockSourceSplit>>(
+                        OPERATOR_NAME,
+                        mockSource,
+                        getNewSourceCoordinatorContext(),
+                        new CoordinatorStoreImpl(),
+                        params,
+                        null) {
+                    @Override
+                    void announceCombinedWatermark() {
+                        counter2.incrementAndGet();
+                    }
+                };
+
+        final int subtask = 0;
+        int attemptNumber = 0;
+        sourceCoordinator.start();
+        setReaderTaskReady(sourceCoordinator, subtask, attemptNumber);
+
+        sourceCoordinator.handleEventFromOperator(
+                subtask, attemptNumber, new ReportedWatermarkEvent(1000));
+
+        Thread.sleep(2000);
```

Comment:
Try to avoid using the sleep inside of all unit tests.

Using the 

- `CountDownLatch countDownLatch = new CountDownLatch(2);`
- `countDownLatch.countDown();`
- `countDownLatch.await();`

instead of 

- `AtomicInteger counter2 = new AtomicInteger(0);`
- `countDownLatch.countDown();`
- `countDownLatch.await();`

 We can check whether the announceCombinedWatermark of  old `sourceCoordinator`  is called after the announceCombinedWatermark of new  `sourceCoordinator`  is called twice.